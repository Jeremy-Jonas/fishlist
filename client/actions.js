import request from 'superagent'

export function getFishPending () {
  return {
    type: 'GET_FISH_PENDING'
  }
}

export function getFishSuccess (fish) {
  return {
    type: 'GET_FISH_SUCCESS',
    fish
  }
}

export function getFishDetailSuccess (fishDetail) {
  return {
    type: 'GET_FISH_DETAIL',
    fishDetail
  }
}

export function getFishError (message) {
  return {
    type: 'GET_FISH_ERROR',
    message
  }
}

export function getFish () {
  return dispatch => {
    dispatch(getFishPending())

    request
      .get('/api/v1/fish')
      .then(res => dispatch(getFishSuccess(res.body)))
      .catch(err => dispatch(getFishError(err.message)))
  }
}

export function getFishDetail (fishId) {
  return dispatch => {
    dispatch(getFishPending())

    request
      .get(`/api/v1/fish/${fishId}`)
      .then(res => dispatch(getFishDetailSuccess(res.body)))
      .catch(err => dispatch(getFishError(err.message)))
  }
}
