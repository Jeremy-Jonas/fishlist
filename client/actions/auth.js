import request from 'superagent'

import { setToken } from '../utils/tokens'

export const loginPending = () => ({ type: 'LOGIN_PENDING' })

export const loginSuccess = () => ({ type: 'LOGIN_SUCCESS' })

export const loginFailure = error => ({ type: 'LOGIN_FAILURE', error })

export const login = (username, password) => dispatch => {
  dispatch(loginPending())

  request
    .post('/api/v1/auth/login')
    .send({ username, password })
    .then(res => {
      setToken(res.body.token)
      dispatch(loginSuccess())
    })
    .catch(err => dispatch(loginFailure(err.response.body.error)))
}

export const registerPending = () => ({ type: 'REGISTER_PENDING' })

export const registerSuccess = () => ({ type: 'REGISTER_SUCCESS' })

export const registerFailure = error => ({ type: 'REGISTER_FAILURE', error })

export const register = (username, password, email) => dispatch => {
  dispatch(registerPending())

  return request
    .post('/api/v1/auth/register')
    .send({ username, password, email })
    .then(res => {
      setToken(res.body.token)
      dispatch(registerSuccess())
    })
    .catch(err => dispatch(registerFailure(err.response.body.error)))
}

export const logout = () => ({ type: 'LOGOUT' })
  
