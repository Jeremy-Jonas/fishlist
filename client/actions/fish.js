import request from 'superagent'

export function getFishPending () {
  return {
    type: 'GET_FISH_PENDING'
  }
}

export function getFishSuccess (fish) {
  return {
    type: 'GET_FISH_SUCCESS',
    fish
  }
}

export function getFishDetailSuccess (fishDetail) {
  return {
    type: 'GET_FISH_DETAIL_SUCCESS',
    fishDetail
  }
}

export function getFishError (message) {
  return {
    type: 'GET_FISH_ERROR',
    message
  }
}

export function orderAZ (sortOrder) {
  return {
    type: 'SORT_AZ',
    sortOrder
  }
}

export function orderZA (sortOrder) {
  return {
    type: 'SORT_ZA',
    sortOrder
  }
}

export function getFish () {
  return dispatch => {
    dispatch(getFishPending())

    request
      .get('/api/v1/fish')
      .then(res => dispatch(getFishSuccess(res.body)))
      .catch(err => dispatch(getFishError(err.message)))
  }
}

export function getFishDetail (fishId) {
  return dispatch => {
    dispatch(getFishPending())

    request
      .get(`/api/v1/fish/${fishId}`)
      .then(res => dispatch(getFishDetailSuccess(res.body)))
      .catch(err => dispatch(getFishError(err.message)))
  }
}

export function searchFish (searchedFish) {
  return {
    type: 'SEARCH_FISH',
    searchedFish
  }
}
