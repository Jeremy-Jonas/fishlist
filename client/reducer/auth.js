const defaultState = {
  error: null,
  loggedIn: false,
  newRegistration: false,
  pending: false
}

export default function authReducer (state = defaultState, { type, error }) {
  switch (type) {
    case 'LOGIN_PENDING':
    case 'REGISTER_PENDING':
      return {
        ...state,
        loggedIn: false,
        pending: true
      }

    case 'LOGIN_SUCCESS':
      return {
        ...state,
        error: null,
        loggedIn: true,
        pending: false
      }

    case 'LOGIN_FAILURE':
    case 'REGISTER_FAILURE':
      return {
        ...state,
        error,
        loggedIn: false,
        pending: false
      }

    case 'REGISTER_SUCCESS':
      return {
        ...state,
        error: null,
        loggedIn: true,
        newRegistration: true,
        pending: false
      }

    default:
      return state
  }
}
