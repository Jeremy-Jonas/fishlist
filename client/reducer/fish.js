export default function fishReducer (state = [], action) {
  switch (action.type) {
    case 'GET_FISH_SUCCESS':
      return action.fish

    default:
      return state
  }
}
