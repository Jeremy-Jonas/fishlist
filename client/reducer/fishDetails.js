export default function fishDetailsReducer (state = {}, action) {
  switch (action.type) {
    case 'GET_FISH_DETAIL_SUCCESS':
      return action.fishDetail

    case 'CLEAR_FISH_DETAIL':
      return {}

    default:
      return state
  }
}
