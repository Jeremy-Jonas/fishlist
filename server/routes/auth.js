const express = require('express')

const router = express.Router()

router.post('/login', validateLogin, checkUser, issueToken)
router.post('/register', validateRegistration, createUser, issueToken)

function checkUser (req, res, next) {
}

function createUser (req, res, next) {
}

function issueToken (req, res, next) {
}

function login (req, res, next) {
}

function register (req, res, next) {
}

function validateLogin (req, res, next) {
}

function validateRegistration (req, res, next) {
}

module.exports = {
  login,
  register
}
