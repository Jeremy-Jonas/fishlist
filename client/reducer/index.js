import { combineReducers } from 'redux'

import auth from './auth'
import fish from './fish'
import info from './info'

export default combineReducers({
  auth,
  fish,
  info
})
