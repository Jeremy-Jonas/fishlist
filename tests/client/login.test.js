import nock from 'nock'

import * as auth from '../../client/actions/auth'
import reducer from '../../client/reducer/auth'

let state = null

beforeAll(() => {
  nock(/.*/)
    .persist()
    .post('/api/v1/auth/login', { username: 'foo', password: 'bar' })
    .reply(201, { ok: true, token: '0123456789ABCDEF' })
    .post('/api/v1/auth/login', { username: 'floof' })
    .reply(401, { ok: false, error: 'No password provided.' })
    .post('/api/v1/auth/register', { username: 'foo', password: 'bar', email: 'bar@foo' })
    .reply(201, { ok: true, token: '0123456789ABCDEF' })
    .post('/api/v1/auth/register', { username: 'floof', password: 'florf' })
    .reply(401, { ok: false, error: 'No email provided.' })
})

afterAll(() => {
  nock.cleanAll()
})

beforeEach(() => {
  state = {
    error: null,
    loggedIn: false,
    newRegistration: false,
    pending: false
  }
})

test('loginPending matches the last snapshot', () => {
  expect(auth.loginPending()).toMatchSnapshot()
})

test('loginSuccess matches the last snapshot', () => {
  expect(auth.loginSuccess()).toMatchSnapshot()
})

test('loginFailure matches the last snapshot', () => {
  expect(auth.loginFailure('badness happened')).toMatchSnapshot()
})

test('login dispatches PENDING', () => {
  expect.assertions(1)
  const expected = { type: 'LOGIN_PENDING' }
  const dispatch = jest.fn()
  return auth.login('foo', 'bar')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('login dispatches SUCCESS', () => {
  expect.assertions(1)
  const expected = { type: 'LOGIN_SUCCESS' }
  const dispatch = jest.fn()
  return auth.login('foo', 'bar')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('login dispatches FAILURE on error', () => {
  expect.assertions(1)
  const expected = { type: 'LOGIN_FAILURE', error: 'No password provided.' }
  const dispatch = jest.fn()
  return auth.login('floof')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('login sets ACCESS-TOKEN correctly', () => {
  expect.assertions(1)
  const expected = '0123456789ABCDEF'
  const dispatch = jest.fn()

  return auth.login('foo', 'bar')(dispatch)
    .then(() => {
      const actual = localStorage.getItem('ACCESS-TOKEN')
      expect(actual).toBe(expected)
    })
})

test('registerPending matches the last snapshot', () => {
  expect(auth.registerPending()).toMatchSnapshot()
})

test('registerSuccess matches the last snapshot', () => {
  expect(auth.registerSuccess()).toMatchSnapshot()
})

test('registerFailure matches the last snapshot', () => {
  expect(auth.registerFailure('Username taken')).toMatchSnapshot()
})

test('register dispatches PENDING', () => {
  expect.assertions(1)
  const expected = { type: 'REGISTER_PENDING' }
  const dispatch = jest.fn()
  return auth.register('foo', 'bar', 'bar@foo')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('register dispatches SUCCESS', () => {
  expect.assertions(1)
  const expected = { type: 'REGISTER_SUCCESS' }
  const dispatch = jest.fn()
  return auth.register('foo', 'bar', 'bar@foo')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('register dispatches FAILURE on error', () => {
  expect.assertions(1)
  const expected = { type: 'REGISTER_FAILURE', error: 'No email provided.' }
  const dispatch = jest.fn()
  return auth.register('floof', 'florf')(dispatch)
    .then(() => expect(dispatch).toHaveBeenCalledWith(expected))
})

test('reducer sets loggedIn true on LOGIN_SUCCESS', () => {
  const expected = {
    loggedIn: true,
    pending: false
  }
  const actual = reducer(state, auth.loginSuccess())
  expect(actual).toMatchObject(expected)
})

test('reducer sets pending true on LOGIN_PENDING', () => {
  const expected = {
    loggedIn: false,
    pending: true
  }
  const actual = reducer(state, auth.loginPending())
  expect(actual).toMatchObject(expected)
})

test('reducer sets error on LOGIN_FAILURE', () => {
  const expected = {
    error: 'Password incorrect.',
    loggedIn: false,
    pending: false
  }
  const actual = reducer(state, auth.loginFailure('Password incorrect.'))
  expect(actual).toMatchObject(expected)
})

test('reducer sets pending true on REGISTER_PENDING', () => {
  const expected = {
    loggedIn: false,
    pending: true
  }
  const actual = reducer(state, auth.registerPending())
  expect(actual).toMatchObject(expected)
})

test('reducer sets error on REGISTER_FAILURE', () => {
  const expected = {
    error: 'Email not provided.',
    loggedIn: false,
    pending: false
  }
  const actual = reducer(state, auth.loginFailure('Email not provided.'))
  expect(actual).toMatchObject(expected)
})

test('reducer sets state correctly on REGISTER_SUCCESS', () => {
  const expected = {
    error: null,
    loggedIn: true,
    newRegistration: true,
    pending: false
  }
  const actual = reducer(state, auth.registerSuccess())
  expect(actual).toMatchObject(expected)
})
