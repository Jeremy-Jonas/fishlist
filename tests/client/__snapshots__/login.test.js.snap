// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginFailure matches the last snapshot 1`] = `
Object {
  "error": "badness happened",
  "type": "LOGIN_FAILURE",
}
`;

exports[`loginPending matches the last snapshot 1`] = `
Object {
  "type": "LOGIN_PENDING",
}
`;

exports[`loginSuccess matches the last snapshot 1`] = `
Object {
  "type": "LOGIN_SUCCESS",
}
`;

exports[`registerFailure matches the last snapshot 1`] = `
Object {
  "error": "Username taken",
  "type": "REGISTER_FAILURE",
}
`;

exports[`registerPending matches the last snapshot 1`] = `
Object {
  "type": "REGISTER_PENDING",
}
`;

exports[`registerSuccess matches the last snapshot 1`] = `
Object {
  "type": "REGISTER_SUCCESS",
}
`;
